File: hex_terrain_editor/model/hex_grid.py
"""HexGrid data model storing the complete map as a dictionary of hex cells."""

from __future__ import annotations

from typing import Iterator

from .hex_cell import HexCell
from .terrain_types import Climate, Elevation
from ..utils.hex_math import hex_neighbors, hex_to_pixel


class HexGrid:
    """Sparse hex grid stored as a dictionary keyed by (q, r) tuples."""

    def __init__(self, width: int = 60, height: int = 30, hex_size: float = 64.0):
        self.width = width
        self.height = height
        self.hex_size = hex_size
        self._cells: dict[tuple[int, int], HexCell] = {}

    def get(self, q: int, r: int) -> HexCell | None:
        """Get the cell at (q, r), or None if it doesn't exist."""
        return self._cells.get((q, r))

    def set(self, q: int, r: int, cell: HexCell) -> None:
        """Set the cell at (q, r)."""
        self._cells[(q, r)] = cell

    def get_or_create(self, q: int, r: int) -> HexCell:
        """Get the cell at (q, r), creating a default one if it doesn't exist."""
        key = (q, r)
        if key not in self._cells:
            self._cells[key] = HexCell(q=q, r=r)
        return self._cells[key]

    def remove(self, q: int, r: int) -> None:
        """Remove the cell at (q, r) if it exists."""
        self._cells.pop((q, r), None)

    def __contains__(self, key: tuple[int, int]) -> bool:
        return key in self._cells

    def __iter__(self) -> Iterator[HexCell]:
        return iter(self._cells.values())

    def __len__(self) -> int:
        return len(self._cells)

    def cells(self) -> Iterator[HexCell]:
        """Iterate over all cells."""
        return iter(self._cells.values())

    def keys(self) -> Iterator[tuple[int, int]]:
        """Iterate over all (q, r) keys."""
        return iter(self._cells.keys())

    def get_neighbors(self, q: int, r: int) -> list[HexCell]:
        """Return existing neighbor cells of (q, r)."""
        result = []
        for nq, nr in hex_neighbors(q, r):
            cell = self.get(nq, nr)
            if cell is not None:
                result.append(cell)
        return result

    def get_neighbor_coords_and_cells(
        self, q: int, r: int
    ) -> list[tuple[int, int, HexCell]]:
        """Return (nq, nr, cell) for existing neighbors."""
        result = []
        for nq, nr in hex_neighbors(q, r):
            cell = self.get(nq, nr)
            if cell is not None:
                result.append((nq, nr, cell))
        return result

    def initialize_rectangular(
        self,
        default_elevation: Elevation = Elevation.PLAINS,
        default_climate: Climate = Climate.Cf,
    ) -> None:
        """Fill the grid with default hex cells in a rectangular layout."""
        for r in range(self.height):
            for q in range(self.width):
                cell = HexCell(
                    q=q,
                    r=r,
                    elevation=default_elevation,
                    climate=default_climate,
                )
                self._cells[(q, r)] = cell

    def bounds(self) -> tuple[int, int, int, int]:
        """Return (min_q, min_r, max_q, max_r) of all cells."""
        if not self._cells:
            return (0, 0, 0, 0)
        qs = [k[0] for k in self._cells]
        rs = [k[1] for k in self._cells]
        return (min(qs), min(rs), max(qs), max(rs))

    def world_bounds(self) -> tuple[float, float, float, float]:
        """Return (min_x, min_y, max_x, max_y) in world pixel coordinates."""
        if not self._cells:
            return (0.0, 0.0, 0.0, 0.0)
        min_x = float("inf")
        min_y = float("inf")
        max_x = float("-inf")
        max_y = float("-inf")
        for q, r in self._cells:
            x, y = hex_to_pixel(q, r, self.hex_size)
            min_x = min(min_x, x)
            min_y = min(min_y, y)
            max_x = max(max_x, x)
            max_y = max(max_y, y)
        # Add margin for hex size
        margin = self.hex_size * 1.5
        return (min_x - margin, min_y - margin, max_x + margin, max_y + margin)

    def to_dict_list(self) -> list[dict]:
        """Serialize all cells to a list of dictionaries."""
        return [cell.to_dict() for cell in self._cells.values()]

    def load_from_dict_list(self, data: list[dict]) -> None:
        """Load cells from a list of dictionaries, clearing existing cells."""
        self._cells.clear()
        for cell_data in data:
            cell = HexCell.from_dict(cell_data)
            self._cells[(cell.q, cell.r)] = cell

File: src/model/hex_grid.rs
use std::collections::HashMap;
use super::hex_cell::HexCell;
use super::terrain_types::{Climate, Elevation};
use crate::hex_math;

pub struct HexGrid {
    pub width: u32,
    pub height: u32,
    pub hex_size: f32,
    pub cells: HashMap<(i32, i32), HexCell>,
    /// Offset applied when mapping axial coords to the data texture.
    /// Set by `build_hex_data_texture`.
    pub grid_offset: (i32, i32),
}

impl HexGrid {
    pub fn new(width: u32, height: u32, hex_size: f32) -> Self {
        Self {
            width,
            height,
            hex_size,
            cells: HashMap::new(),
            grid_offset: (0, 0),
        }
    }

    pub fn get(&self, q: i32, r: i32) -> Option<&HexCell> {
        self.cells.get(&(q, r))
    }

    pub fn get_mut(&mut self, q: i32, r: i32) -> Option<&mut HexCell> {
        self.cells.get_mut(&(q, r))
    }

    pub fn set(&mut self, cell: HexCell) {
        self.cells.insert((cell.q, cell.r), cell);
    }

    pub fn remove(&mut self, q: i32, r: i32) -> Option<HexCell> {
        self.cells.remove(&(q, r))
    }

    /// Fill a rectangular region of the grid with default cells.
    pub fn initialize_rectangular(&mut self, default_elevation: Elevation, default_climate: Climate) {
        self.cells.clear();
        for r in 0..self.height as i32 {
            for q in 0..self.width as i32 {
                let mut cell = HexCell::new(q, r);
                cell.elevation = default_elevation;
                cell.climate = default_climate;
                self.cells.insert((q, r), cell);
            }
        }
    }

    /// Iterate over all cell coordinates.
    pub fn cells(&self) -> impl Iterator<Item = (i32, i32)> + '_ {
        self.cells.keys().copied()
    }

    /// Iterate over all cells.
    pub fn cells_iter(&self) -> impl Iterator<Item = &HexCell> {
        self.cells.values()
    }

    /// Compute the bounding box in axial coordinates.
    pub fn bounds(&self) -> (i32, i32, i32, i32) {
        if self.cells.is_empty() {
            return (0, 0, 0, 0);
        }
        let mut min_q = i32::MAX;
        let mut min_r = i32::MAX;
        let mut max_q = i32::MIN;
        let mut max_r = i32::MIN;
        for &(q, r) in self.cells.keys() {
            min_q = min_q.min(q);
            min_r = min_r.min(r);
            max_q = max_q.max(q);
            max_r = max_r.max(r);
        }
        (min_q, min_r, max_q, max_r)
    }

    /// Compute the bounding box in world (pixel) coordinates, including hex
    /// extents at the edges.
    pub fn world_bounds(&self, hex_size: f32) -> (f32, f32, f32, f32) {
        if self.cells.is_empty() {
            return (0.0, 0.0, 0.0, 0.0);
        }
        let mut min_x = f32::MAX;
        let mut min_y = f32::MAX;
        let mut max_x = f32::MIN;
        let mut max_y = f32::MIN;
        let hw = hex_math::hex_width(hex_size) * 0.5;
        let hh = hex_math::hex_height(hex_size) * 0.5;
        for &(q, r) in self.cells.keys() {
            let (px, py) = hex_math::hex_to_pixel(q, r, hex_size);
            min_x = min_x.min(px - hw);
            min_y = min_y.min(py - hh);
            max_x = max_x.max(px + hw);
            max_y = max_y.max(py + hh);
        }
        (min_x, min_y, max_x, max_y)
    }

    /// Return references to all existing neighbor cells for a given hex.
    pub fn get_neighbors(&self, q: i32, r: i32) -> Vec<&HexCell> {
        let neighbors = hex_math::hex_neighbors(q, r);
        neighbors
            .iter()
            .filter_map(|&(nq, nr)| self.cells.get(&(nq, nr)))
            .collect()
    }

    /// Build a data texture for GPU upload.
    ///
    /// Each pixel at position `(q - min_q, r - min_r)` stores:
    /// - R: texture_index
    /// - G: elevation value (discriminant)
    /// - B: 0
    /// - A: 255
    ///
    /// Returns `(pixel_data, texture_width, texture_height)`.
    /// Also updates `self.grid_offset` with `(min_q, min_r)`.
    pub fn build_hex_data_texture(&mut self) -> (Vec<u8>, u32, u32) {
        let (min_q, min_r, max_q, max_r) = self.bounds();
        self.grid_offset = (min_q, min_r);

        let tex_w = (max_q - min_q + 1).max(1) as u32;
        let tex_h = (max_r - min_r + 1).max(1) as u32;
        let mut data = vec![0u8; (tex_w * tex_h * 4) as usize];

        for cell in self.cells.values() {
            let col = (cell.q - min_q) as u32;
            let row = (cell.r - min_r) as u32;
            let idx = ((row * tex_w + col) * 4) as usize;
            if idx + 3 < data.len() {
                data[idx] = cell.texture_index() as u8;
                data[idx + 1] = cell.elevation as u8;
                data[idx + 2] = 0;
                data[idx + 3] = 255;
            }
        }

        (data, tex_w, tex_h)
    }
}
